confmap/confmap.go

```diff
@@ -21,9 +22,12 @@ import (
 
 const (
 	// KeyDelimiter is used as the default key delimiter in the default koanf instance.
-	KeyDelimiter = "::"
+	KeyDelimiter      = "::"
+	EnvVarNamePattern = `^[a-zA-Z_][a-zA-Z0-9_]*$`
 )
 
+var EnvVarNameRegexp = regexp.MustCompile(EnvVarNamePattern)
```

Comment:
I'd rather move this to a `confmap/internal` package, because
- upstream may change the way this is expressed (see open-telemetry/opentelemetry-specification/issues/4027)
- morally, this feels like something that should be exposed by the `go.opentelemetry.io/contrib/config` module instead of us